### ISSUE TITLE
Possible change to QueryProcessor.get_iterable

### ISSUE THREAD

**Issue by murphyke** [chop-dbhi/avocado#273]
I noticed while developing a custom `QueryProcessor` that the `get_queryset` method is called twice while servicing /api/data/preview/.

I was wondering if it would make sense to modify [get_iterable](https://github.com/chop-dbhi/avocado/blob/master/avocado/query/pipeline.py#L46-L65) to take an optional `queryset` parameter so that if the caller already had the queryset at hand, it could be reused.

The context is that in Serrano's [preview.py](https://github.com/chop-dbhi/serrano/blob/master/serrano/resources/preview.py#L46-L55), there is an explicit call to grab the queryset, but a chance for reuse is missed when `get_iterable` is called.

This change would relieve custom query processors from having to manage caching themselves if they perform heavy computations.

**Comment by bruth** [comment]
Yes. The `QueryProcessor` class is confusing and honestly was rushed at the time it was designed. For a little background, the original intent of the query processor was to:

1. Encapsulate the boilerplate logic for preparing a query and the exporter
2. Provide a way to customize the processor that can be changed at runtime

The encapsulation attempts to hide some of the side effects from the translation of a Django queryset, which is object-based, to a *Harvest query* which is row-based. The transformation between these models actually happens when the `DataView` [is applied](https://github.com/chop-dbhi/avocado/blob/master/avocado/query/oldparsers/dataview.py#L144-L149) so fields anywhere in the data model (not just on the queryset's model) can be added to the query. The `queryset.select` call in the view code ultimately calls [this modeltree method](https://github.com/chop-dbhi/modeltree/blob/master/modeltree/tree.py#L934-L969) which changes the internal `queryset.query` object. The processor ultimately [only uses](https://github.com/chop-dbhi/avocado/blob/master/avocado/query/pipeline.py#L62-L63) the internal query object since the raw query is executed.

The customization aspect is solved by the [query processor registry](https://github.com/chop-dbhi/avocado/blob/master/avocado/query/pipeline.py#L100). Multiple processors can be defined by name and selected at runtime. Although a processor *could* be a function, I chose it to be class-based with small methods so individual parts can be extended.

As you pointed out, the poor encapsulation leaks out and is confusing and I think this is an appropriate opportunity to change the implementation before the 2.4 release. Does the above make sense? How might we change this so it feels more natural? /cc @tjrivera

**Comment by naegelyd** [comment]
@murphyke Technically, the `get_iterable` does accept a queryset keyword argument and that is passed to `get_queryset` but I understand what you are saying about a missed chance for reuse. I believe modifying `get_iterable` to just use the supplied queryset would modify the expected behavior of the method as it stands now. I think that the `get_queryset` result could be cached(for later reuse in get_iterable), or, there could be a method added(utility method or tack it on to the QuerySet) to convert an existing queryset to an iterable without considering the context or view. 

Basically, I imagine a new method(`to_iterable` for the sake of this conversation) that would take a `queryset`, `offset`, and `limit` and essentially run [this code](https://github.com/chop-dbhi/avocado/blob/master/avocado/query/pipeline.py#L50-L63) to convert the queryset to an iterable. 

Under the above proposition, the `preview` endpoint(and anyone else who has a vested interest) could do the following:

```python
# Build a queryset for pagination and other downstream use
queryset = processor.get_queryset(request=request)

...

iterable = to_iterable(queryset, offset, limit)
```

@bruth Like you said, this portion of the code can be confusing so anything that can be done to trim it down and make it clearer should definitely be part of the 2.4 release IMO.

**Comment by bruth** [comment]
Thanks @naegelyd. After writing the explanation above, my gut is that much of this will resolve itself if effort is put into distinguishing the Django vs. Harvest boundary for a query. My assumption all these years was that Harvest would allow you to "bring your own queryset" and pass into the Harvest query processor. However, in practice this never appears to be necessary. That is, Harvest is strictly for ad-hoc queries, not a replacement or even an alternate to constructing queries the way Django allows you to using the ORM. This is because if you know what you want, you just build the queryset explicitly.. which is no different from writing an explicit SQL query for BI-type reporting.

Simply put, if a context or view is used, Harvest will take over the query and do whatever needs to be done to execute it. Ideally a Harvest query would stand alone independent of a Django queryset. There would be a converter from a queryset to the Harvest query.. possibly by extracting the structure of the query and converting it into context and view objects. The fact that (right now) the query processor uses the internals of the Django ORM is independent of need to interop with querysets. I say this because I am thinking of future applications as well where the Django ORM is no longer the only "query execution path".

**Comment by bruth** [comment]
This issue has been alleviated in 07a27eb for 2.3.x and #301 for 2.4.x. The `get_iterable` uses the passed `queryset` directly otherwise it builds one using `get_queryset`. In addition, the logic has been fixed Serrano to pass the queryset into appropriate calls in a9fc02d for 2.3.x and #266 for 2.4.x.
